plugins/auth-backend/src/providers/bitbucket/provider.ts

```diff
@@ -0,0 +1,316 @@
+      refreshToken: privateInfo.refreshToken,
+    };
+  }
+
+  async refresh(req: OAuthRefreshRequest): Promise<OAuthResponse> {
+    const { accessToken, params } = await executeRefreshTokenStrategy(
+      this._strategy,
+      req.refreshToken,
+      req.scope,
+    );
+    const fullProfile = await executeFetchUserProfileStrategy(
+      this._strategy,
+      accessToken,
+    );
+    return this.handleResult({
+      fullProfile,
+      params,
+      accessToken,
+      refreshToken: req.refreshToken,
+    });
+  }
+
+  private async handleResult(result: BitbucketOAuthResult) {
+    result.fullProfile.avatarUrl =
+      result.fullProfile._json!.links!.avatar!.href;
+    const { profile } = await this.authHandler(result);
+
+    const response: OAuthResponse = {
+      providerInfo: {
+        idToken: result.params.id_token,
+        accessToken: result.accessToken,
+        scope: result.params.scope,
+        expiresInSeconds: result.params.expires_in,
+      },
+      profile,
+    };
+
+    if (this.signInResolver) {
+      response.backstageIdentity = await this.signInResolver(
+        {
+          result,
+          profile,
+        },
+        {
+          tokenIssuer: this.tokenIssuer,
+          catalogIdentityClient: this.catalogIdentityClient,
+          logger: this.logger,
+        },
+      );
+    }
+
+    return response;
+  }
+}
+
+export const bitbucketUsernameSignInResolver: SignInResolver<BitbucketOAuthResult> =
+  async (info, ctx) => {
+    const { result } = info;
+
+    if (!result.fullProfile.username) {
+      throw new Error('Bitbucket profile contained no Username');
+    }
+
+    const entity = await ctx.catalogIdentityClient.findUser({
+      annotations: {
+        'bitbucket.org/username': result.fullProfile.username,
+      },
+    });
+
+    const claims = getEntityClaims(entity);
+    const token = await ctx.tokenIssuer.issueToken({ claims });
+
+    return { id: entity.metadata.name, entity, token };
+  };
+
+export const bitbucketUserIdSignInResolver: SignInResolver<BitbucketOAuthResult> =
+  async (info, ctx) => {
+    const { result } = info;
+
+    if (!result.fullProfile.id) {
+      throw new Error('Bitbucket profile contained no User ID');
+    }
+
+    const entity = await ctx.catalogIdentityClient.findUser({
+      annotations: {
+        'bitbucket.org/user-id': result.fullProfile.id,
+      },
+    });
+
+    const claims = getEntityClaims(entity);
+    const token = await ctx.tokenIssuer.issueToken({ claims });
+
+    return { id: entity.metadata.name, entity, token };
+  };
+
+export type BitbucketProviderOptions = {
+  /**
+   * The profile transformation function used to verify and convert the auth response
+   * into the profile that will be presented to the user.
+   */
+  authHandler?: AuthHandler<OAuthResult>;
+
+  /**
+   * Configure sign-in for this provider, without it the provider can not be used to sign users in.
+   */
+  signIn: {
```

Comment:
```suggestion
  signIn?: {
```

One last thing 😅

Reply:
@Rugvip I thought the whole point of removing default is to make a choice of resolver mandatory, otherwise how would the catalog findUser know what annotation to look for?

Reply:
So without a sign-in resolver you wouldn't be able to use the provider to sign-in to Backstage at all. More concretely there won't be any `BackstageIdentity` in the response. The provider will still be usable to delegate access to resources, so you can still for example use it to get a Bitbucket token that you can then use to talk to the Bitbucket API.

Reply:
hi @Rugvip it doesn't work. I've applied above changes. When signInResolver is undefined, something is still trying to access backstageIdentity.Id from the return from handleResult() call.
With this set, it gets weird type error by the way:
```
      const signInResolver: SignInResolver<BitbucketOAuthResult> | undefined =
        options?.signIn?.resolver
          ? (info =>
            options?.signIn?.resolver(info, {
              catalogIdentityClient,
              tokenIssuer,
              logger,
            }))
          : undefined;
```
Above is checking and returning undefined if resolver option is left unset. 
I can verify that this.signInResolver is indeed undefined inside handleResult and so setting of backstage identify is skipped  and yet this is the outcome below.

I'll attempt again to fix it tomorrow but meanwhile if it's obvious to you please feel free to make that final change. Thank you for your help!
![image](https://user-images.githubusercontent.com/30108240/135915563-27f9dd8b-f6ea-4703-a1ee-241db5e7e39b.png)

Reply:
@Rugvip restoring minimal form of default resolver fixes it, as long as it returns {id: userId} example:
```
      const defaultResolver: SignInResolver<BitbucketOAuthResult> =
        async (info, ctx) => {
          const { fullProfile } = info.result;

          const userId = fullProfile.username || fullProfile.id;

          const token = await ctx.tokenIssuer.issueToken({
            claims: { sub: userId, ent: [`user:default/${userId}`] },
          });

          return { id: userId, token };
        };

      const signInResolverFn =
        options?.signIn?.resolver ?? defaultResolver;
```

This, as I understand, would still allow for:

> you can still for example use it to get a Bitbucket token that you can then use to talk to the Bitbucket API

and for users choosing Resolvers in factories.ts. 
Does that sound acceptable?